doc/adr/0005-wego-core-auth-strategy.md

```diff
@@ -0,0 +1,128 @@
+
+These deploy keys are outside the scope of a git repo, so we will need a higher level of access in the form of a Git Provider access token to add deploy keys programmatically.
+
+### Git Provider
+
+Operations such as creating pull requests (or "merge requests" in the case of Gitlab), adding deploy keys, and retrieving user data will need to be completed via HTTP requests. Each major Git Provider has some form of OAuth flow:
+
+- [GitHub OAuth Flow](https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#web-application-flow)
+  - Notice that `localhost` redirect_uris are permitted: https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#localhost-redirect-urls
+  - This approach requires a `CLIENT_SECRET` be embedded in the binary, thus exposing the secret leaking by decompiling the binary
+- [Github Device Flow](https://docs.github.com/en/developers/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps#device-flow)
+  - This approach doesn't require a `CLIENT_SECRET` or callback page
+  - We can do these steps from a browser if we feel that is the best UX
+- [Gitlab Proof Key for Code Exchange (PKCE) Flow](https://docs.gitlab.com/ee/api/oauth2.html#authorization-code-with-proof-key-for-code-exchange-pkce)
+  - Note that the documentation explicitly calls out that PKCE is optimal for client-side apps without a public cloud server (our use case)
+  - The documentation example specifies a `CLIENT_SECRET`, but that may be a documentation bug. A `CLIENT_SECRET` should not be necessary
+- [Bitbucket Implicit Grant](https://developer.atlassian.com/cloud/bitbucket/oauth-2/#2--implicit-grant--4-2-)
+  - No `CLIENT_SECRET` required here
+
+### Kubernetes Cluster
+
+All operations will need to go through an API server that is installed on the cluster via the `wego gitops install` command. This API server will run with a service account that provides the ability to do the following:
+
+- Create `Namespace` resources
+- Create `Application`, `Source`, and `Kustomize|Helm` resources
+- Read Kubernetes objects in the `wego-system` namespace
+
+### Security Concerns
+
+Each of the Git Providers listed above support some sort of "personal access token" that allow for "impersonation" style authorization. These tokens never expire, and provide "sudo" access to the user's Github account.
+
+For this reason, we want to utilize the more short-lived OAuth tokens to avoid exposing a personal access token. We also don't want to persist these tokens longer than we have to.
+
+In addition, flows that require a `CLIENT_SECRET` will require us to embed it in the binary, which leaves us open to exposing it via decompilation.
+
```

Comment:
```suggestion
CI will inject any sensitive information into the binary to avoid keeping data in version control in a public repo. We are still susceptible to leaking data via decompilation.
```